File: NostrVault/app/src/main/java/com/nostrvault/data/local/ProfileRepository.kt
package com.nostrvault.data.local

import android.content.Context
import com.nostrvault.data.model.FeedProfile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.io.File

/**
 * Port of ProfileRepository.swift -- profile and relay-list cache persistence.
 * Pure JSON file I/O with no UI dependencies.
 */
object ProfileRepository {

    private val json = Json { ignoreUnknownKeys = true; prettyPrint = false }
    private var dataDir: File? = null

    val isInitialized: Boolean get() = dataDir != null

    fun init(context: Context) {
        dataDir = File(context.filesDir, "nostrvault_data").also { it.mkdirs() }
    }

    private val profilesFile: File? get() = dataDir?.let { File(it, "profiles.json") }
    private val relayListsFile: File? get() = dataDir?.let { File(it, "relay_lists.json") }
    private val outboxRelaysFile: File? get() = dataDir?.let { File(it, "outbox_relays.json") }
    private val dmRelayListsFile: File? get() = dataDir?.let { File(it, "dm_relay_lists.json") }
    private val serverListsFile: File? get() = dataDir?.let { File(it, "server_lists.json") }

    // ---- Load from disk ----

    fun loadProfiles(): Map<String, FeedProfile> = profilesFile?.let { loadMap(it) } ?: emptyMap()

    fun loadRelayLists(): Map<String, List<String>> = relayListsFile?.let { loadStringListMap(it) } ?: emptyMap()

    fun loadOutboxRelays(): Map<String, List<String>> = outboxRelaysFile?.let { loadStringListMap(it) } ?: emptyMap()

    fun loadDMRelayLists(): Map<String, List<String>> = dmRelayListsFile?.let { loadStringListMap(it) } ?: emptyMap()

    fun loadServerLists(): Map<String, List<String>> = serverListsFile?.let { loadStringListMap(it) } ?: emptyMap()

    // ---- Save to disk (background) ----

    suspend fun saveProfiles(profiles: Map<String, FeedProfile>) = withContext(Dispatchers.IO) {
        saveToFile(profilesFile ?: return@withContext, json.encodeToString(profiles))
    }

    suspend fun saveRelayLists(lists: Map<String, List<String>>) = withContext(Dispatchers.IO) {
        saveToFile(relayListsFile ?: return@withContext, json.encodeToString(lists))
    }

    suspend fun saveOutboxRelays(lists: Map<String, List<String>>) = withContext(Dispatchers.IO) {
        saveToFile(outboxRelaysFile ?: return@withContext, json.encodeToString(lists))
    }

    suspend fun saveDMRelayLists(lists: Map<String, List<String>>) = withContext(Dispatchers.IO) {
        saveToFile(dmRelayListsFile ?: return@withContext, json.encodeToString(lists))
    }

    suspend fun saveServerLists(lists: Map<String, List<String>>) = withContext(Dispatchers.IO) {
        saveToFile(serverListsFile ?: return@withContext, json.encodeToString(lists))
    }

    // ---- Event parsing (pure functions) ----

    /**
     * Parse Kind 0 (Metadata) event content into profile field updates.
     * Returns the updated profile and whether any field changed, or null if invalid JSON.
     */
    fun parseMetadataContent(
        content: String,
        pubkey: String,
        existingProfile: FeedProfile?,
    ): Pair<FeedProfile, Boolean>? {
        val metadata = try {
            json.parseToJsonElement(content) as? JsonObject ?: return null
        } catch (_: Exception) {
            return null
        }

        val name = metadata["name"]?.jsonPrimitive?.contentOrNull
        val displayName = metadata["display_name"]?.jsonPrimitive?.contentOrNull
        val picture = metadata["picture"]?.jsonPrimitive?.contentOrNull
        val nip05 = metadata["nip05"]?.jsonPrimitive?.contentOrNull
        val about = metadata["about"]?.jsonPrimitive?.contentOrNull
        val lud16 = metadata["lud16"]?.jsonPrimitive?.contentOrNull
        val lud06 = metadata["lud06"]?.jsonPrimitive?.contentOrNull
        val website = metadata["website"]?.jsonPrimitive?.contentOrNull

        val existing = existingProfile ?: FeedProfile(pubkey = pubkey)
        var changed = false

        val profile = existing.copy(
            name = name.also { if (it != existing.name) changed = true },
            displayName = displayName.also { if (it != existing.displayName) changed = true },
            pictureURL = picture.also { if (it != existing.pictureURL) changed = true },
            nip05 = nip05.also { if (it != existing.nip05) changed = true },
            about = about.also { if (it != existing.about) changed = true },
            lud16 = lud16.also { if (it != existing.lud16) changed = true },
            lud06 = lud06.also { if (it != existing.lud06) changed = true },
            website = website.also { if (it != existing.website) changed = true },
        )

        return profile to changed
    }

    /**
     * Parse Kind 10002 (NIP-65 Relay List Metadata) tags.
     * Tags: ["r", relay_url, "read" | "write"], no marker = both.
     */
    fun parseRelayListTags(tags: List<List<String>>): Pair<List<String>, List<String>> {
        val inbox = mutableListOf<String>()
        val write = mutableListOf<String>()
        for (tag in tags) {
            if (tag.size < 2 || tag[0] != "r") continue
            val url = tag[1]
            if (!isValidRelayUrl(url)) continue
            val marker = if (tag.size >= 3) tag[2] else ""
            if (marker == "read" || marker == "") inbox.add(url)
            if (marker == "write" || marker == "") write.add(url)
        }
        return inbox to write
    }

    /** Parse Kind 10050 (NIP-17 DM Relay List) tags. */
    fun parseDMRelayListTags(tags: List<List<String>>): List<String> =
        tags.filter { it.size >= 2 && it[0] == "r" && isValidRelayUrl(it[1]) }.map { it[1] }

    /** Parse Kind 10063 (BUD-03 User Server List) tags. */
    fun parseServerListTags(tags: List<List<String>>): List<String> =
        tags.filter { it.size >= 2 && it[0] == "server" }.map { it[1] }

    /** Reject relay URLs with spaces or missing wss:// scheme. */
    private fun isValidRelayUrl(url: String): Boolean =
        url.startsWith("wss://") && !url.contains(' ')

    /** Parse Kind 10000 (Mute List) p-tags into blocked hex pubkeys. */
    fun parseMuteListPTags(tags: List<List<String>>): List<String> =
        tags.filter { it.size >= 2 && it[0] == "p" }.map { it[1] }

    // ---- Private helpers ----

    private inline fun <reified T> loadMap(file: File): Map<String, T> = try {
        if (file.exists()) json.decodeFromString(file.readText()) else emptyMap()
    } catch (_: Exception) {
        emptyMap()
    }

    private fun loadStringListMap(file: File): Map<String, List<String>> = try {
        if (file.exists()) json.decodeFromString(file.readText()) else emptyMap()
    } catch (_: Exception) {
        emptyMap()
    }

    private fun saveToFile(file: File, content: String) {
        try {
            dataDir?.mkdirs()
            file.writeText(content)
        } catch (_: Exception) { }
    }
}

File: HavenApp/HavenApp/Services/ProfileRepository.swift
import Foundation

/// Pure profile and relay-list cache operations — load, save, and parse.
/// No Combine, no @Published, no SwiftUI. All file I/O is explicit.
/// Direct translation target for Kotlin (Room DB or DataStore).
enum ProfileRepository {

    // MARK: - File Paths

    private static func havenSupportDir() -> URL {
        guard let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else {
            return URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("Haven")
        }
        let havenDir = appSupport.appendingPathComponent("Haven", isDirectory: true)
        try? FileManager.default.createDirectory(at: havenDir, withIntermediateDirectories: true)
        return havenDir
    }

    static var profilesURL: URL { havenSupportDir().appendingPathComponent("profiles.json") }
    static var relayListsURL: URL { havenSupportDir().appendingPathComponent("relay_lists.json") }
    static var outboxRelaysURL: URL { havenSupportDir().appendingPathComponent("outbox_relays.json") }
    static var dmRelayListsURL: URL { havenSupportDir().appendingPathComponent("dm_relay_lists.json") }
    static var serverListsURL: URL { havenSupportDir().appendingPathComponent("server_lists.json") }

    // MARK: - Load from Disk

    static func loadProfiles() -> [String: FeedProfile] {
        guard let data = try? Data(contentsOf: profilesURL),
              let loaded = try? JSONDecoder().decode([String: FeedProfile].self, from: data) else { return [:] }
        return loaded
    }

    static func loadRelayLists() -> [String: [String]] {
        guard let data = try? Data(contentsOf: relayListsURL),
              let loaded = try? JSONDecoder().decode([String: [String]].self, from: data) else { return [:] }
        return loaded
    }

    static func loadOutboxRelays() -> [String: [String]] {
        guard let data = try? Data(contentsOf: outboxRelaysURL),
              let loaded = try? JSONDecoder().decode([String: [String]].self, from: data) else { return [:] }
        return loaded
    }

    static func loadDMRelayLists() -> [String: [String]] {
        guard let data = try? Data(contentsOf: dmRelayListsURL),
              let loaded = try? JSONDecoder().decode([String: [String]].self, from: data) else { return [:] }
        return loaded
    }

    static func loadServerLists() -> [String: [String]] {
        guard let data = try? Data(contentsOf: serverListsURL),
              let loaded = try? JSONDecoder().decode([String: [String]].self, from: data) else { return [:] }
        return loaded
    }

    // MARK: - Save to Disk (background queue)

    static func saveProfiles(_ profiles: [String: FeedProfile]) {
        let copy = profiles
        DispatchQueue.global(qos: .utility).async {
            let dir = havenSupportDir()
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            if let data = try? JSONEncoder().encode(copy) {
                try? data.write(to: profilesURL)
            }
        }
    }

    static func saveRelayLists(_ lists: [String: [String]]) {
        let copy = lists
        DispatchQueue.global(qos: .utility).async {
            if let data = try? JSONEncoder().encode(copy) { try? data.write(to: relayListsURL) }
        }
    }

    static func saveOutboxRelays(_ lists: [String: [String]]) {
        let copy = lists
        DispatchQueue.global(qos: .utility).async {
            if let data = try? JSONEncoder().encode(copy) { try? data.write(to: outboxRelaysURL) }
        }
    }

    static func saveDMRelayLists(_ lists: [String: [String]]) {
        let copy = lists
        DispatchQueue.global(qos: .utility).async {
            if let data = try? JSONEncoder().encode(copy) { try? data.write(to: dmRelayListsURL) }
        }
    }

    static func saveServerLists(_ lists: [String: [String]]) {
        let copy = lists
        DispatchQueue.global(qos: .utility).async {
            if let data = try? JSONEncoder().encode(copy) { try? data.write(to: serverListsURL) }
        }
    }

    // MARK: - Event Parsing (pure functions)

    /// Parses Kind 0 (Metadata) event content into profile field updates.
    /// Returns the updated profile and whether any field actually changed,
    /// or nil if the content isn't valid JSON.
    static func parseMetadataContent(
        _ content: String,
        pubkey: String,
        existingProfile: FeedProfile?
    ) -> (profile: FeedProfile, changed: Bool)? {
        guard let metadata = try? JSONSerialization.jsonObject(
            with: content.data(using: .utf8) ?? Data()
        ) as? [String: Any] else { return nil }

        let name = metadata["name"] as? String
        let displayName = metadata["display_name"] as? String
        let picture = (metadata["picture"] as? String).flatMap { URL(string: $0) }
        let nip05 = metadata["nip05"] as? String
        let about = metadata["about"] as? String
        let lud16 = metadata["lud16"] as? String
        let lud06 = metadata["lud06"] as? String
        let website = metadata["website"] as? String

        var profile = existingProfile ?? FeedProfile(pubkey: pubkey)
        var changed = false

        if profile.name != name { profile.name = name; changed = true }
        if profile.displayName != displayName { profile.displayName = displayName; changed = true }
        if profile.pictureURL != picture { profile.pictureURL = picture; changed = true }
        if profile.nip05 != nip05 { profile.nip05 = nip05; changed = true }
        if profile.about != about { profile.about = about; changed = true }
        if profile.lud16 != lud16 { profile.lud16 = lud16; changed = true }
        if profile.lud06 != lud06 { profile.lud06 = lud06; changed = true }
        if profile.website != website { profile.website = website; changed = true }

        return (profile, changed)
    }

    /// Parses Kind 10002 (Relay List Metadata / NIP-65) tags into inbox and write relay lists.
    /// Tags: ["r", relay_url, "read" | "write"], no marker = both.
    static func parseRelayListTags(_ tags: [[String]]) -> (inbox: [String], write: [String]) {
        var inbox: [String] = []
        var write: [String] = []
        for tag in tags {
            guard tag.count >= 2, tag[0] == "r" else { continue }
            let marker = tag.count >= 3 ? tag[2] : ""
            if marker == "read" || marker == "" { inbox.append(tag[1]) }
            if marker == "write" || marker == "" { write.append(tag[1]) }
        }
        return (inbox, write)
    }

    /// Parses Kind 10050 (NIP-17 DM Relay List) tags into relay URLs.
    /// Tags: ["r", relay_url].
    static func parseDMRelayListTags(_ tags: [[String]]) -> [String] {
        tags.compactMap { tag in
            tag.count >= 2 && tag[0] == "r" ? tag[1] : nil
        }
    }

    /// Parses Kind 10063 (BUD-03 User Server List) tags into Blossom server URLs.
    /// Tags: ["server", url].
    static func parseServerListTags(_ tags: [[String]]) -> [String] {
        tags.compactMap { tag in
            tag.count >= 2 && tag[0] == "server" ? tag[1] : nil
        }
    }

    /// Parses Kind 10000 (Mute List) p-tags into blocked hex pubkeys.
    static func parseMuteListPTags(_ tags: [[String]]) -> [String] {
        tags.compactMap { tag in
            tag.count >= 2 && tag[0] == "p" ? tag[1] : nil
        }
    }

    /// Converts hex pubkeys to npub strings for mute list storage.
    static func hexKeysToNpubs(_ hexKeys: [String]) -> [String] {
        hexKeys.compactMap { hexKey in
            guard let data = Bech32.hexToData(hexKey) else { return nil }
            return Bech32.encode(hrp: "npub", data: data)
        }
    }
}
